integration-tests/tests/src/tests/serialization.ts

```diff
@@ -484,7 +484,8 @@ describe("JSON serialization", () => {
       afterEach(() => {
         if (realm) {
           realm.write(() => {
-            realm.deleteAll();
+            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+            realm!.deleteAll();
```

Comment:
I'm surprised this is needed ...

Reply:
Yeah I have been confused by this before. I believe the logic is that TS can't infer how the callback is called, so it could be being called async and `realm` has changed in the interim to being `null`

![image](https://user-images.githubusercontent.com/5458070/148383580-ee8bf664-a9aa-4d9d-ae54-4558a2af2fdf.png)

Reply:
But `realm` is captured when the lambda is created? It can't change value from that point on 🤔

Reply:
It seems like a TSC bug to me.

Reply:
Hmm yeah... maybe the TS logic is a bit faulty, I think this is described in https://github.com/microsoft/TypeScript/issues/10982#issuecomment-247993896

Reply:
It works as expected here: https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEYD2A7AzgF3nKEC2AXPCgK4QTwA+ABQN7wBmSSRARAEayvwC+A3ACgBASwbwAFDnwBKeLQHxFiVJkbN4AXmwhceAHRMkgpcvRZoedsCiaJsjQD45Ck0uTokEEHohIA5pI6+AbM0sZK-AI8AkA

Reply:
I hope its not

Reply:
That's weird...

Reply:
Ahh. Perhaps https://github.com/realm/realm-js/pull/4193/files#diff-13ec3f38b0cf2f5f7842047684a331ad82a406f144a3e9ac84cad52718bd824eR491 is the offender :)

Reply:
I think it's actually because `realm` is a `let` https://github.com/realm/realm-js/pull/4193/files#diff-13ec3f38b0cf2f5f7842047684a331ad82a406f144a3e9ac84cad52718bd824eR475

See https://www.typescriptlang.org/play?ssl=2&ssc=1&pln=2&pc=4#code/PTAEBMFMGMBsEMBOlTQPYDsDOAXUz5YBbALlAwFdZZQAfAAoG9QAzNNMgIgCMlPQAvgG4AULEh4CxMpWp0mrdl16J+A0AF5QzNh1A8+g0SICWLUAAopRAJTaRoR6ky5FaTfkiEiAOl2inZ2w8BCJucHgPCzsNAD57QMD0bDRxH1g0AHMrL2I-dhsAp2ERAREgA

Reply:
Oh my ... I learned something today.